DataFormats/L1THGCal/interface/HGCalClusterT.h

```diff
@@ -40,67 +42,132 @@ namespace l1t
+
+            double cMipt = c->mipPt()*fraction;
+
+            if( constituents_.empty() )
+            { 
+                detId_ = HGCalDetId( c->detId() );
+                seedMipPt_ = cMipt;
+              /* if the centre will not be dynamically calculated 
+                 the seed centre is considere as cluster centre */
+              if( !updateCentre )
+              {
+              Basic3DVector<float> constituentCentre( c->position() );
+              centre_ = GlobalPoint( constituentCentre ); 
+          }
         }
-        centre_ = GlobalPoint( clusterCentre );
 
-        if( clusterCentre.z()!=0 ) 
-        {
-          centreProj_= GlobalPoint( clusterCentre / clusterCentre.z() );
+        /* update cluster positions (IF requested) */
+        if( updateCentre ){
+            Basic3DVector<float> constituentCentre( c->position() );
+            Basic3DVector<float> clusterCentre( centre_ );
+            
+            clusterCentre = clusterCentre*mipPt_ + constituentCentre*cMipt;
+            if( (mipPt_ + cMipt ) > 0 ) 
+            {
+                clusterCentre /= ( mipPt_ +  cMipt );
+            }
+            centre_ = GlobalPoint( clusterCentre );
+            
+            if( clusterCentre.z()!=0 ) 
+            {
+                centreProj_= GlobalPoint( clusterCentre / clusterCentre.z() );
+            }
         }
+
         /* update cluster energies */
-        mipPt_ += c->mipPt();
+        mipPt_ += cMipt;
 
-        int updatedPt = hwPt() + c->hwPt();
-        setHwPt(updatedPt);
+        int updatedPt = hwPt() + (int)(c->hwPt()*fraction);
+        setHwPt( updatedPt );
 
         math::PtEtaPhiMLorentzVector updatedP4 ( p4() );
-        updatedP4 += c->p4(); 
+        updatedP4 += (c->p4()*fraction); 
         setP4( updatedP4 );
 
         constituents_.push_back( c );
+        constituentsV_.push_back( c );
+        constituentsFraction_.push_back( fraction );
 
       }
       
-      bool valid() const { return valid_;}
-      void setValid(bool valid) { valid_ = valid;}
+      void removeConstituent( const edm::Ptr<C>& c, bool updateCentre=true ){
+          
+          /* remove the pointer to c from the edm::PtrVector */
+          double fraction=0;
+          for( unsigned i=0; i<constituentsV_.size(); i++ )
+          {
+              if( constituentsV_[i] == c )
+              {
+                  // remove constituent and get its fraction in the cluster
+                  constituentsV_.erase( constituentsV_.begin()+i );
+                  fraction = constituentsFraction_.at(i);
+                  constituentsFraction_.erase( constituentsFraction_.begin()+i );
+                  break;
+              }
+          }
+          
+          /* rebuilding the constituents_ persistent vector */
+          this->fillPersistentPtrVector_(); 
+
+          /* update cluster positions (IF requested) */
+          double cMipt = c->mipPt()*fraction;
```

Comment:
If the constituent has not been found and there is nothing to remove, I think the following code can be skipped.